### PR TITLE
Add the `ignoreAnnotated` array parameter to the FunctionNaming rule

### PR DESCRIPTION
**Pull request by gtcompscientist**
This is intended to ease the use of Jetpack Compose.

Addresses #2733

### CODE REVIEW

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/naming/FunctionNamingSpec.kt [hunk 1]
```diff
@@ -87,5 +87,30 @@ class FunctionNamingSpec : Spek({
...
+        """
+            assertThat(FunctionNaming().compileAndLint(code)).hasSourceLocations(
+                SourceLocation(2, 18),
+                SourceLocation(6, 18)
+            )
+            val config = TestConfig(mapOf(FunctionNaming.IGNORE_ANNOTATED to arrayOf("Suppress")))
+            assertThat(FunctionNaming(config).compileAndLint(code)).hasSourceLocations(
+                SourceLocation(2, 18),
+                SourceLocation(10, 19)
+            )
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/naming/FunctionNamingSpec.kt:113]: This should be in a separate test.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/naming/FunctionNaming.kt [hunk 2]
```diff
@@ -36,6 +38,15 @@ class FunctionNaming(config: Config = Config.empty) : Rule(config) {
     private val functionPattern by LazyRegex(FUNCTION_PATTERN, "([a-z][a-zA-Z0-9]*)|(`.*`)")
     private val excludeClassPattern by LazyRegex(EXCLUDE_CLASS_PATTERN, "$^")
     private val ignoreOverridden = valueOrDefault(IGNORE_OVERRIDDEN, true)
+    private val ignoreAnnotated = valueOrDefaultCommaSeparated(IGNORE_ANNOTATED, emptyList())
+            .map { it.removePrefix("*").removeSuffix("*") }
+
+    private lateinit var annotationExcluder: AnnotationExcluder
```
Comment (cortinico) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/naming/FunctionNaming.kt:44]: You should use this field before reporting (like add it to the condition on line 58/60)

Comment (gtcompscientist) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/naming/FunctionNaming.kt:45]: Done

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/naming/FunctionNamingSpec.kt (hunk 1 above)
Comment (gtcompscientist) [reply detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/naming/FunctionNamingSpec.kt:129]: Separated the two ways into two tests.

**Comment by gtcompscientist** [comment]
Updated the PR based on comments received.

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/naming/FunctionNamingSpec.kt [hunk 3]
```diff
@@ -87,5 +87,46 @@ class FunctionNamingSpec : Spek({
...
+                @Composable
+                fun SHOULD_BE_FLAGGED() {}
+            }
+        """
+            val config = TestConfig(mapOf(FunctionNaming.IGNORE_ANNOTATED to arrayOf("Suppress")))
+            assertThat(FunctionNaming(config).compileAndLint(code)).hasSourceLocations(
+                SourceLocation(2, 18),
+                SourceLocation(10, 19)
+            )
+        }
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/naming/FunctionNamingSpec.kt:130]: The following avoids code duplication.

```suggestion
        describe("annotated functions") {
            val code = """
                class D {
                    fun SHOULD_BE_FLAGGED() {}
                }
                class E {
                    @Suppress
                    fun SHOULD_BE_FLAGGED() {}
                }
                class F {
                    @Composable
                    fun SHOULDNT_BE_FLAGGED() {}
                }

            """

            it("Ignores default annotated functions") {
                assertThat(FunctionNaming().compileAndLint(code)).hasSourceLocations(
                    SourceLocation(2, 18),
                    SourceLocation(6, 18)
                )
            }
    
            it("Ignores annotated functions if ignoreAnnotated includes the given annotation class") {
                val config = TestConfig(mapOf(FunctionNaming.IGNORE_ANNOTATED to arrayOf("Suppress")))
                assertThat(FunctionNaming(config).compileAndLint(code)).hasSourceLocations(
                    SourceLocation(2, 18),
                    SourceLocation(10, 19)
                )
            }
        }
```

**Comment by schalkms** [comment]
java.lang.ClassCastException: class [Ljava.lang.String; cannot be cast to class java.util.List ([Ljava.lang.String; and java.util.List are in module java.base of loader 'bootstrap')

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/naming/FunctionNaming.kt [hunk 4]
```diff
@@ -36,6 +40,14 @@ class FunctionNaming(config: Config = Config.empty) : Rule(config) {
     private val functionPattern by LazyRegex(FUNCTION_PATTERN, "([a-z][a-zA-Z0-9]*)|(`.*`)")
     private val excludeClassPattern by LazyRegex(EXCLUDE_CLASS_PATTERN, "$^")
     private val ignoreOverridden = valueOrDefault(IGNORE_OVERRIDDEN, true)
+    private val ignoreAnnotated = valueOrDefault(IGNORE_ANNOTATED, listOf("Composable"))
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/naming/FunctionNaming.kt:43]:
```suggestion
    private val ignoreAnnotated = valueOrDefaultCommaSeparated(IGNORE_ANNOTATED, listOf("Composable"))
```

Comment (gtcompscientist) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/naming/FunctionNaming.kt]: Updated and squashed.

**Comment by schalkms** [comment]
Have you executed the tests locally?

```
NamingConventionLengthSpec > should not report underscore variable names FAILED
47810
    kotlin.UninitializedPropertyAccessException: lateinit property annotationExcluder has not been initialized
```

**Comment by gtcompscientist** [comment]
> NamingConventionLengthSpec

I was running them in the wrong location. 👎 

I believe I have followed the directions more closely this time around. And I got a successful build after publishing locally.

**Comment by schalkms** [comment]
We use the Kotlin Jsr223 mechanism to ensure valid Kotlin test code snippets. That’s the reason why the build was failing. Thanks again for contributing! Cool stuff..

**Pull request merged**